File: day5/p1/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

type ConversionRange struct {
	source_start int
	source_end   int
	target_start int
	target_end   int
	translate    int
}

func convert(c *ConversionRange, num int) int {
	if num >= c.source_start && num <= c.source_end {
		return num + c.translate
	} else {
		return -1
	}
}

func print(c ConversionRange) {
	fmt.Println()
}

func main() {
	file, _ := os.Open("./test.text")
	defer file.Close()

	scanner := bufio.NewScanner(file)
	var seed_to_soil []*ConversionRange

	// Parse file
	index := 0
	s_to_s_mapping := false
	for scanner.Scan() {
		line := scanner.Text()

		// Get starting seeds

		if index == 0 {
			seeds_list := strings.Split(line, ": ")[1]
			seeds_value_list := strings.Split(seeds_list, " ")

			var seeds []int

			for _, seed := range seeds_value_list {
				seed_value, _ := strconv.Atoi(seed)
				seeds = append(seeds, seed_value)
			}

			fmt.Println("Seeds: ", seeds)
		}

		if line == "seed-to-soil map:" {
			s_to_s_mapping = true
			continue
		}

		if s_to_s_mapping {
			if line == "" {
				s_to_s_mapping = false
				for _, r := range seed_to_soil {
					print(r)
				}
				continue
			} else {
				var r_values []int
				range_strings := strings.Split(line, " ")
				for _, r := range range_strings {
					r_value, _ := strconv.Atoi(r)
					r_values = append(r_values, r_value)
				}

				var temp *ConversionRange
				temp = new(ConversionRange)

				temp.target_start = r_values[0]
				temp.source_start = r_values[1]
				length := r_values[2]

				temp.target_end = temp.target_start + length
				temp.source_end = temp.source_start + length
				temp.translate = temp.target_start - temp.target_end

				seed_to_soil = append(seed_to_soil, temp)
			}
		}

		index++
	}

}

File: day5/p1/main.py
#!/bin/python3.12

FILE_PATH = "input.text"

class ConversionRange:
    target_start : int
    target_end : int
    source_start : int
    source_end : int
    translate : int

    def __init__(self, target_start : int, source_start : int, length : int):
        self.target_start = target_start
        self.source_start = source_start
        self.target_end = self.target_start + length
        self.source_end = self.source_start + length
        self.translate = self.target_start - self.source_start

    def __str__(self):
        print(f"Target start: {self.target_start}")
        print(f"Source start: {self.source_start}")
        print(f"Translate: {self.translate}")
        return ""



class Conversion:
    ranges : list[ConversionRange]

    def __init__(self):
        self.ranges = []
        pass

    def convert(self, num : int) -> int:
        for r in self.ranges:
            if num >= r.source_start and num <= r.source_end:
                return num + r.translate
        return num


def main() -> None:
    with open(FILE_PATH, 'r') as f:
        contents = f.read().split('\n')

    # Setup
    seeds: list[int] = []
    maps: dict[str, Conversion] = {}
    maps["seed-to-soil"] = Conversion()
    maps["soil-to-fertilizer"] = Conversion()
    maps["fertilizer-to-water"] = Conversion()
    maps["water-to-light"] = Conversion()
    maps["light-to-temperature"] = Conversion()
    maps["temperature-to-humidity"] = Conversion()
    maps["humidity-to-location"] = Conversion()

    current_conversion: str = ""

    # Parse the file
    for index, line in enumerate(contents):
        if index == 0:
            seeds_list: list[str] = line.split(": ")[1].split()
            for seed in seeds_list:
                seeds.append(int(seed))
            print(f"Seeds: {seeds}")
            continue

        if not line:
            continue

        if not line[0].isdigit():
            current_conversion = line[:-5]
            print(f"Current conversion: {current_conversion}")
            continue

        nums: list[int] = list(map(lambda x: int(x), line.split()))

        temp = ConversionRange(nums[0], nums[1], nums[2])
        print(temp)
        maps[current_conversion].ranges.append(temp)

    # Convert the seeds
    lowest_location: float = float('inf')
    for seed in seeds:
        temp = seed
        for key, value in maps.items():
            print(f"Mapping {temp} with {key}")
            temp = value.convert(temp)

        if temp < lowest_location:
            lowest_location = temp
        print(f"Location: {temp}")

    print(f"Lowest location: {lowest_location}")




if __name__ == "__main__":
    main()



